core/src/no/group15/playmagic/ecs/GameMap.kt

```diff
@@ -130,66 +109,102 @@ class GameMap (
 	}
 
 
+	// --- IMPLEMENTATION ---
+
+
+	private fun overlapping(rectangle: Rectangle, tile: MatrixIndexes): Boolean {
+		val tile = toWorldCoordinate(tile)
+		val tileBoundingBox = Rectangle()
+		tileBoundingBox.height = 1f
+		tileBoundingBox.width = 1f
+		tileBoundingBox.setCenter(Vector2(tile.x, tile.y))
+
+		return rectangle.overlaps(tileBoundingBox)
+	}
+
+
+	private fun isRigidTile(tile: MatrixIndexes): Boolean {
+		return mapMatrix[tile.y][tile.x] != CellType.EMPTY
+	}
+
 	/**
 	 * Returns the map coordinate of the tiles around the
 	 * entity. If the entity is the one marked with 'e' in the diagram
 	 *     t t t
 	 *     t e t
 	 *     t t t
 	 * the function return all the eight tiles marked with 't'
-	 * plus the one marked with 'e'
+	 * plus the one marked with 'e'.
 	 */
-	private fun nearTiles(entity: Entity): Iterable<Coordinate> {
-		val list = LinkedList<Coordinate>()
-		val center = Vector2()
-		val entityPoint = boundingBox(entity).getCenter(center)
-		for (i in max(0, entityPoint.x.toInt() - 1) until min(entityPoint.x.toInt() + 1, mapMatrix[0].size - 1) + 1) {
-			for (j in max(0, entityPoint.y.toInt() - 1) until min(entityPoint.y.toInt() + 1, mapMatrix.size - 1) + 1) {
-				list.add(Coordinate(i.toFloat(), j.toFloat()))
+	private fun nearTiles(tile: MatrixIndexes): Iterable<MatrixIndexes> {
+		val list = LinkedList<MatrixIndexes>()
+
+		for (i in max(0, tile.x - 1)..min(tile.x + 1, width() - 1)) {
+			for (j in max(0, tile.y - 1)..min(tile.y + 1, height() - 1)) {
+				list.add(MatrixIndexes(i, j))
 			}
 		}
+
 		return list
 	}
 
 
+	/**
+	 * Translate the world coordinate to the matrix indexes.
+	 */
+	private fun toMatrixIndexes(c: WorldCoordinate): MatrixIndexes {
+		var xFloor = floor(c.x).toInt()
+		var yFloor = floor(c.y).toInt()
 
-	private fun boundingBox (entity: Entity): Rectangle {
-		// TODO replace this with existing boundingbox in transform component
-		val transform = entity.getComponent(TransformComponent::class.java)
-		val texture = entity.getComponent(TextureComponent::class.java)
-		val coordinate = toMatrixIndexes(Coordinate(transform.position.x, transform.position.y))
-
-		return Rectangle(
-			coordinate.x,
-			coordinate.y,
-			transform.boundingBox.width * transform.scale.x,
-			transform.boundingBox.width * transform.scale.y
-		)
+		return MatrixIndexes(xFloor + width() / 2, (height() - 1) / 2 - yFloor)
 	}
 
 
 	/**
-	 * Translate the world coordinate to the matrix indexes.
+	 * Translate the matrix indexes to the world coordinate.
 	 */
-	private fun toMatrixIndexes(coordinate: Coordinate): Coordinate {
-		return Coordinate(coordinate.x - topLeftCorner.x, topLeftCorner.y - coordinate.y)
+	private fun toWorldCoordinate(m: MatrixIndexes): WorldCoordinate {
+		return WorldCoordinate(m.x - width() / 2 + .5f, (height() - 1) / 2 - m.y + .5f)
```

Comment:
One line functions expressions like this one can be declared `fun ... = <expr>` instead of `fun ... { return <expr> }`